src/courtroom.cpp

```diff
@@ -4008,6 +4030,157 @@ void Courtroom::append_ic_text(QString p_text, QString p_name, QString p_action,
+    if (log_newline)
+      // For some reason, we're forced to use <br> instead of the more sensible
+      // \n. Why? Because \n is treated as a new Block instead of a soft newline
+      // within a paragraph!
+      ui_ic_chatlog_pinned->textCursor().insertHtml("<br>");
+    else
+      ui_ic_chatlog_pinned->textCursor().insertText(": ", normal);
+    // Format the result in italics
+    ui_ic_chatlog->textCursor().insertText(p_text + ".", italics);
+  }
+  else {
+    if (log_newline) {
+      // For some reason, we're forced to use <br> instead of the more sensible
+      // \n. Why? Because \n is treated as a new Block instead of a soft newline
+      // within a paragraph!
+      ui_ic_chatlog_pinned->textCursor().insertHtml("<br>");
+    }
+    else {
+      ui_ic_chatlog_pinned->textCursor().insertText(": ", normal);
+    }
+    // Format the result according to html
+    if (log_colors) {
+      QString p_text_filtered = filter_ic_text(p_text, true, -1, color);
+      p_text_filtered =
+          p_text_filtered.replace("$c0", chatlog_color.name(QColor::HexArgb));
+      for (int c = 1; c < max_colors; ++c) {
+        QColor color_result = default_color_rgb_list.at(c);
+        p_text_filtered = p_text_filtered.replace(
+            "$c" + QString::number(c), color_result.name(QColor::HexArgb));
+      }
+      ui_ic_chatlog_pinned->textCursor().insertHtml(p_text_filtered);
+    }
+    else {
+      ui_ic_chatlog_pinned->textCursor().insertText(filter_ic_text(p_text, false),
+                                             normal);
+    }
+  }
+
+  // Only append with newline if log goes upwards
+  if (!log_goes_downwards && need_newline) {
+    ui_ic_chatlog_pinned->textCursor().insertBlock(format);
+  }
+
+  // If we got too many blocks in the current log, delete some.
+  while (ui_ic_chatlog_pinned->document()->blockCount() > log_maximum_blocks &&
+         log_maximum_blocks > 0) {
+    QTextCursor temp_curs = ui_ic_chatlog_pinned->textCursor();
+    temp_curs.movePosition(log_goes_downwards ? QTextCursor::Start
+                                              : QTextCursor::End);
+    temp_curs.select(QTextCursor::BlockUnderCursor);
+    temp_curs.removeSelectedText();
+    if (log_goes_downwards)
+      temp_curs.deleteChar();
+    else
+      temp_curs.deletePreviousChar();
+  }
+
+  // Finally, scroll the scrollbar to the correct position.
+  if (old_cursor.hasSelection() ||
+      old_scrollbar_value != scrollbar_target_value) {
+    // The user has selected text or scrolled away from the bottom: maintain
+    // position.
+    ui_ic_chatlog_pinned->setTextCursor(old_cursor);
+    ui_ic_chatlog_pinned->verticalScrollBar()->setValue(old_scrollbar_value);
+  }
+  else {
+    ui_ic_chatlog_pinned->verticalScrollBar()->setValue(
+        log_goes_downwards ? ui_ic_chatlog->verticalScrollBar()->maximum() : 0);
+  }
+}
+
+void Courtroom::clear_pinned_message() { ui_ic_chatlog_pinned->clear(); }
+
```

Comment:
# 🚨 COPY PASTE ALERT 🚨
please revise, never copy-paste this much code! Adjust the existing function(s) instead and refactor them to allow your use case.